packages/make-styles-webpack-loader/src/webpackLoader.ts

```diff
@@ -0,0 +1,89 @@
+import { EvalCache, Module } from '@linaria/babel';
+import * as enhancedResolve from 'enhanced-resolve';
+import * as path from 'path';
+import * as webpack from 'webpack';
+
+import { transformSync, TransformResult, TransformOptions } from './transformSync';
+
+type WebpackLoaderParams = Parameters<webpack.LoaderDefinitionFunction<never>>;
+
+/**
+ * Webpack can also pass sourcemaps as a string, Babel accepts only objects.
+ * See https://github.com/babel/babel-loader/pull/889.
+ */
+function parseSourceMap(inputSourceMap: WebpackLoaderParams[1]): TransformOptions['inputSourceMap'] {
+  try {
+    if (typeof inputSourceMap === 'string') {
+      return JSON.parse(inputSourceMap) as TransformOptions['inputSourceMap'];
+    }
+
+    return inputSourceMap as TransformOptions['inputSourceMap'];
+  } catch (err) {
+    return undefined;
+  }
+}
+
+export function webpackLoader(
+  this: webpack.LoaderContext<never>,
+  sourceCode: WebpackLoaderParams[0],
+  inputSourceMap: WebpackLoaderParams[1],
+) {
+  EvalCache.clearForFile(this.resourcePath);
+
+  const resolveOptionsDefaults: webpack.ResolveOptions = {
+    conditionNames: ['require'],
+    extensions: ['.js', '.jsx', '.ts', '.tsx', '.json'],
+  };
+  // ⚠ "this._compilation" is a deprecated API, however there seems to be no other way to access Webpack's resolver
+  // There is this.resolve, but it's asynchronous. Another option is to read the webpack.config.js, but it won't work
+  // for programmatic usage. This API is used by many loaders/plugins, so hope we're safe for a while
+  const resolveOptionsFromWebpackConfig: webpack.ResolveOptions = this._compilation?.options.resolve || {};
+
+  const resolveSync = enhancedResolve.create.sync({
```

Comment:
why is a third party resolver used and not the webpack loader's resolver ? https://webpack.js.org/api/loaders/#thisgetresolve

Reply:
`Module._resolveFilename` is sync while `this.resolve` is async API:

```tsx
resolve(context: string, request: string, callback: function(err, result: string))
```

Under hood Webpack uses the same module, it's part of Webpack's org: https://github.com/webpack/enhanced-resolve

Reply:
Just curious, is there a reason things need to be sync in the loader ?

Reply:
> Just curious, is there a reason things need to be sync in the loader ?

- Linaria uses own implementation of `require` for modules evaluation (it's required for proper module resolution and evaluation of dependencies): https://github.com/callstack/linaria/blob/86486e631d1bfc0fde199be6165159ee9af8a41d/packages/babel/src/module.ts#L179-L184
- `require.resolve()` is a sync API, https://nodejs.org/api/modules.html#modules_require_resolve_request_options
- `Module._resolveFilename` is used in `require.resolve`, that's why it should be sync https://github.com/callstack/linaria/blob/86486e631d1bfc0fde199be6165159ee9af8a41d/packages/babel/src/module.ts#L172